Permissioned token (#292)
scale-it/algo-builder: examples/permissioned-token/README.md

# Permissioned token

In permissioned tokens carry additional requirements and permission mechanism on who, when and how can use that tokens. Securities may use the permissioned token mechanism to define complex compliance requirements, such as investor whitelisting and transfer restrictions. For example, only a verified, KYCed users can exchange tokens.

## Permissioned Token Motivation
Specifically, for Permissioned Tokens, we want to:
- Control the number of tokens minted for distribution and also allow flexibility for the administrator to issue more tokens in the future.
- Create whitelists of individuals who are approved to buy/hold the tokens and segregated sublists, as needed, to support transfer restrictions (e.g., U.S. vs international).

> - Manage lock-ups as needed to ensure there are not transfers before allowed.

Author: yes

- Process requirements or needs that come up post-issuance, such as adding or removing investors to the whitelists or creating additional whitelists for new restrictions.
- Take confidence in the ability to remove restrictions in the future if transfers become open or unlimited.
The permissioned token Template outlines a design of permissioned tokens with emphasis for shares and equity.

## Design Overview

We want to enforce permissioned checks on all token transfer between non-reserve accounts. We will use a stateless TEAL as a clawback, which will execute all transfers. So if user A wants to transfer to user B, (s)he will call a clawback stateless smart contract to transfer from account A to account B.
We implement a Permissioned Token template, which executes the following steps:

1. Create ASA
2. Distribute to initial users

> 3. Freeze all assets OR create the asset default-frozen (clawback can still move frozen tokens).

Author: I think *reminding* should be *remaining*. Also, what do you mean by remaining assets? Should this be move "some/all" assets?

> 3. Freeze all assets OR create the asset default-frozen (clawback can still move frozen tokens).

Reviewer: correct, sorry.

> 3. Freeze all assets OR create the asset default-frozen (clawback can still move frozen tokens).

Reviewer: I was thinking that we will do an initial distribution before we will freez assets and change the clawback address.

> 3. Freeze all assets OR create the asset default-frozen (clawback can still move frozen tokens).

Author: ah, okay

### Specification

Smart contracts (in `/assets`):
- *clawback.py*: The asset clawback is a contract account represent by this smart contract. It's logic ensures that during token transfer between non-reserve accounts, `controller` smart contract is called. It also handles issuance transaction & update reserve account transaction.
- *controller.py*: This smart contract acts as a controller of the token. It ensures that during a token transfer smart contract `P` (with rules and permissions) is called. If there are multiple rules contract, then controller must ensure that all those contracts are called (it must store the app ids). Controller smart contract can also be used to kill the token.
- *permissions.py*: This is the rules smart contract. In this template we have two rules:
  - An account cannot hold more than 100 tokens
  - Both `from` and `to` accounts must be whitelisted
- *clear_state_program.py*: clears app state (returns 1)

`assets/asa.yaml` - This the the asset definition file.

### Setup

Deployment scripts are placed directly in `/scripts`:

> **NOTE:** In a real world scenario, transactions involving a multisig account (eg. token can be created and issued by a multisig) will involve an interaction of many users. A user will receive a signed transaction i.e a signed `deployASA` tx by for a multisig accounts. User can append his/her signature using `algob sign-multisg` command & then use `executeSignedTxFromFile` function to successfully send transaction to network (eg. deploy token).

Reviewer 2: ```suggestion
**NOTE:** In a real world scenario, transactions involving a multisig account (eg. token can be created and issued by a multisig) will involve an interaction of many users. A user will receive a signed transaction/create an unsigned transaction i.e a `deployASA` tx. User can sign the transaction using `algob sign-multisig` command or by using the `signMultiSig` function  & then use the `executeSignedTxFromFile` function to successfully send transaction to network (eg. deploy token).
```

7. *Change Permissions Manager* (`scripts/permissions/change-manager.js`): Token manager can change the permissions manager via an application call to the controller smart contract. If the parameters are correct, tx is accepted and controller smart contract updates the `permission_manager` in it's global state.

### Other Solution

The solution in this template extends [Algorand Dev Office Hour](https://register.gotowebinar.com/recording/recordingView?webinarKey=1651582324861824270&registrantEmail=ratikjindal21%40gmail.com) idea presented by Jason - using asset clawback as escrow for permissioned token.
In this template we add a new contract - `controller.py` which essentially "controls" the asset/token. This controls the general properties of the asset and also ensures that rule(s) smart contract(s) are called.

### Links

Developer hour links:
- *video:* [https://www.youtube.com/watch?v=aMDZamxtR14](https://www.youtube.com/watch?v=aMDZamxtR14)

> - *article:* [https://developer.algorand.org/solutions/assets-and-custom-transfer-logic](https://developer.algorand.org/solutions/assets-and-custom-transfer-logic/)

Author: line 91 doesn't seem correct. Updating to:
```md
- Developer Office Hours | Assets and Custom Transfer Logic Using Algorand Smart Contracts: [video](https://www.youtube.com/watch?v=aMDZamxtR14)
- [Assets and Custom Transfer Logic](https://developer.algorand.org/solutions/assets-and-custom-transfer-logic/)
```

> - *article:* [https://developer.algorand.org/solutions/assets-and-custom-transfer-logic](https://developer.algorand.org/solutions/assets-and-custom-transfer-logic/)

Reviewer: correct, thanks.
